p2pnetwork/node.py

```diff
@@ -75,114 +65,98 @@ def __init__(self, host, port, id=None, callback=None, max_connections=0):
-            the id's of the node. First we send our id and then we receive the id of the node we are connected to.
-            When the connection is made the method outbound_node_connected is invoked. If reconnect is True, the
-            node will try to reconnect to the code whenever the node connection was closed. The method returns
-            True when the node is connected with the specific host."""
+    def connect_with_node(self, host: str, port: int, reconnect: bool = False) -> bool:
+        """Make a connection with another node that is running on host with port.
+
+        When the connection is made, an event is triggered outbound_node_connected. When the connection is made with
+        the node, it exchanges the id's of the node. First we send our id and then we receive the id of the node we
+        are connected to. When the connection is made the method outbound_node_connected is invoked. If reconnect is
+        True, the node will try to reconnect to the code whenever the node connection was closed. The method returns
+        True when the node is connected with the specific host."""
 
         if host == self.host and port == self.port:
             print("connect_with_node: Cannot connect with yourself!!")
             return False
 
         # Check if node is already connected with this node!
-        for node in self.nodes_outbound:
+        for node in self.all_nodes:
             if node.host == host and node.port == port:
-                print("connect_with_node: Already connected with this node (" + node.id + ").")
+                print(f"connect_with_node: Already connected with this node ({node.id}).")
                 return True
 
+        node_ids = [node.id for node in self.all_nodes]
+
         try:
             sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            self.debug_print("connecting to %s port %s" % (host, port))
+            self.debug_print(f"connecting to {host} port {port}")
             sock.connect((host, port))
 
             # Basic information exchange (not secure) of the id's of the nodes!
-            sock.send(self.id.encode('utf-8')) # Send my id to the connected node!
-            connected_node_id = sock.recv(4096).decode('utf-8') # When a node is connected, it sends its id!
+            sock.send(self.id.encode('utf-8'))  # Send my id to the connected node!
+            connected_node_id = sock.recv(4096).decode('utf-8')  # When a node is connected, it sends its id!
 
             # Cannot connect with yourself
-            if self.id == connected_node_id:
-                print("connect_with_node: You cannot connect with yourself?!")
+            if self.id == connected_node_id or connected_node_id in node_ids:
                 sock.send("CLOSING: Already having a connection together".encode('utf-8'))
                 sock.close()
                 return True
 
-            # Fix bug: Cannot connect with nodes that are already connected with us!
```

Comment:
This bug has not been resolved by the code above. Inbound connections cannot be checked by their port. We should check for a similar node id to check if we are already connected to the node.